Add FLIP for invalidating references to transferred resources (#1043)
onflow/flow: flips/20220708-resource-reference-invalidation.md

# Invalidate References to Transferred Resources

| Status        | Proposed                                     |
:-------------- |:---------------------------------------------|
| **FLIP #**    | N/A                                          |
| **Author(s)** | Supun Setunga (supun.setunga@dapperlabs.com) |
| **Sponsor**   | Supun Setunga (supun.setunga@dapperlabs.com) |
| **Updated**   | 2022-07-08                                   |

## Objective

> This proposal is to introduce an invalidation for ephemeral reference values, if the referring resource is transferred

Reviewer: ```suggestion
This proposal is to introduce an invalidation for ephemeral reference values, if the referred-to resource is transferred
```

> This proposal is to introduce an invalidation for ephemeral reference values, if the referring resource is transferred

Reviewer 2: Maybe mention here and also in the title that this proposal is for Cadence. 

Also, it might not be clear to users what transfer means (it is not described in the documentation), so maybe explain it in 1-2 lines.

```cadence
// Create a resource.
let r <-create R()

// And take a reference.
let ref = &r as &R

// Then transfer the resource into an account.
account.save(<-r, to: /storage/r)

// Update the reference.
// This will also update the referring resource in the account.
ref.id = 2
```

In the above example, resource `r` was initially created on the stack and was not owned by anyone.
But similarly, it is also possible for someone to take a reference to a resource in their account, and transfer the
resource to someone, and keep accessing the transferred resource through the reference.
```cadence
// Create a resource collection in 'foo' account
fooAccount.save(<-[<-create R()], to: /storage/a)

// Take a ephemeral reference to the resource at index '0' in the collection.
// Resource is in the 'foo' account at the time of taking the reference.
let collectionRef = fooAccount.borrow<&[R]>(from: /storage/a)!
let ref = &collectionRef[0] as &R

// Transfer the entire collection to the 'bar' account.
let collection <- fooAccount.load<@[R]>(from: /storage/a)!
barAccount.save(<- collection, to: /storage/b)

// Can still access the transferred resource through the reference.
ref.id = 2
```

> ## User Benefit

Reviewer 2: Maybe add an example for the potential security issue here. For example, an analogy, like keeping a reference to an resource like an NFT, transferring the NFT from one user A to another user B, then user A still being able to use the NFT, potentially even mutating it.

Author: I have added a very similar example under the `Motivation` section. Or are you suggesting using a more real-world example, like NFT?

This proposal would avoid a potential foot-gun of gaining/giving/retaining unintended access to resources through
references.

> 

Reviewer 2: Maybe also add an "Implementation Benefit" ("Technical Benefit"?) section mentioning that the current implementation for this feature adds overhead due to book keeping, which would go away once this proposal is accepted/implemented

Author: Good idea. Added in e5130c1b8ca3d6624d0ac49b2dee3a606eb857bc

## Design Proposal
The proposed solution in this FLIP is to invalidate a reference, if the underlying resource is transferred from its
original location where the reference was taken.
There can be different types of transfers, depending on the origin (where the reference was obtained), and the
destination.

| Origin  | Destination       | Reference validity |
|:--------|:------------------|:-------------------|   
| Stack   | Stack             | Valid              |
| Stack   | Account           | Invalid            |
| Account | Stack             | Invalid            |
| Account | Same Account      | Valid              |
| Account | Different Account | Invalid            |

With this change, trying to use variable `ref` in the above two examples will produce a run-time error,
causing the program to terminate.
```cadence
ref.id = 2    // Will produce a run-time error
```

### Drawbacks

> - This is a breaking change, and developers will have to update their codes. Many existing contracts maybe affected.

Reviewer: ```suggestion
- This is a breaking change, and developers will have to update their codes. Many existing contracts may be affected.
```

that the references to resources point to the correct resource, when the referred-to resource value is moved around.
With this proposal, such bookkeeping would no longer be required and the implementation complexity could be eliminated.

## Design Proposal
The proposed solution in this FLIP is to invalidate a reference, if the underlying resource is transferred from its
original location where the reference was taken.
There can be different types of transfers, depending on the origin (where the reference was obtained), and the
destination.

| Origin  | Destination       | Reference validity |
|:--------|:------------------|:-------------------|   
| Stack   | Stack             | Valid              |
| Stack   | Account           | Invalid            |
| Account | Stack             | Invalid            |
| Account | Same Account      | Valid              |
| Account | Different Account | Invalid            |

With this change, trying to use variable `ref` in the above two examples will produce a run-time error,
causing the program to terminate.
```cadence
ref.id = 2    // Will produce a run-time error
```

### Drawbacks

> - This is a breaking change, and developers will have to update their codes. Many existing contracts may be affected.

Reviewer 3: ```suggestion
- This is a breaking change, and developers will have to update their code. Many existing contracts may be affected.
```

references.

## Implementation Benefit
The current implementation of Cadence resource reference tracking requires a lot of bookkeeping in order to ensure
that the references to resources point to the correct resource, when the referred-to resource value is moved around.
With this proposal, such bookkeeping would no longer be required and the implementation complexity could be eliminated.

## Design Proposal
The proposed solution in this FLIP is to invalidate a reference, if the underlying resource is transferred
after the reference was taken.
The reference is invalidated upon the first transfer, regardless of the origin and the destination.

There can be different types of transfers, depending on the origin (where the reference was obtained), and the
destination.

| Origin  | Destination       | Reference validity |
|:--------|:------------------|:-------------------|
| Stack   | Stack             | Invalid            |
| Stack   | Account           | Invalid            |
| Account | Stack             | Invalid            |
| Account | Same Account      | Invalid            |

> | Account | Different Account | Invalid            |

Reviewer 2: Given that all cases are now invalid, maybe remove the third column, and maybe just make this a list
